shared/lib_financial.cpp

```diff
@@ -35,153 +35,240 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+        deriv_sum = irr_derivative_sum(initial_guess, cf_vector, count);
+        if (deriv_sum != 0.0)
+            calculated_irr = calculated_irr - irr_poly_sum(calculated_irr, cf_vector, count) / deriv_sum;
+        else
+            break;
+
+        number_of_iterations++;
+        residual = irr_poly_sum(calculated_irr, cf_vector, count) / scale_factor;
     }
-*/
-	int numberOfIterations=0;
-	double calculatedIRR=0;
-	double initialGuess = 0.1; // 10% is default used in Excel IRR function
+    return calculated_irr;
+}
+
+double libfin::irr(const std::vector<double>& cf_vector, int count, double initial_guess, double tolerance, int max_iterations)
+{
+    int number_of_iterations = 0;
+    double calculated_irr = std::numeric_limits<double>::quiet_NaN();
 
-	if (CashFlows.size() < 3)
-		return initialGuess;
+    if (count < 1)
+        return calculated_irr;
 
-    if ( (Count > 1) && (CashFlows[0] <= 0))
+    // only possible for first value negative
+    if ((cf_vector.at(0) <= 0))
     {
-		double deriv_sum = irr_derivative_sum(initialGuess,CashFlows,Count);
-		if (deriv_sum != 0)
-			calculatedIRR = initialGuess - irr_poly_sum(initialGuess,CashFlows,Count)/deriv_sum;
-		else
-			return initialGuess;
-
-		numberOfIterations++;
-		while (!(std::abs(irr_poly_sum(calculatedIRR,CashFlows,Count)) <= tolerance) && (numberOfIterations < maxIterations))
-		{
-			deriv_sum = irr_derivative_sum(initialGuess,CashFlows,Count);
-			if (deriv_sum != 0.0)
-				calculatedIRR = calculatedIRR - irr_poly_sum(calculatedIRR,CashFlows,Count)/deriv_sum;
-			else
-				break;
-
-			numberOfIterations++;
-		}
-	}
-    return calculatedIRR;
-}
-
-
-/*ported directly from Delphi simple geometric sum*/
-double libfin::npv(double Rate, const std::vector<double> &CashFlows, int Count) //, PaymentTime: TPaymentTime)
-{
-	/*
-{ Caution: The sign of NPV is reversed from what would be expected for standard
-   cash flows!}
-*/
-	if (Rate <= -1.0)
-	{
-		// TODO - throw exception?
-		return -999;
-	}
+        // initial guess from http://zainco.blogspot.com/2008/08/internal-rate-of-return-using-newton.html
+        if ((initial_guess < -1) && (count > 1)) // second order
+        {
+            if (cf_vector.at(0) != 0)
+            {
+                double b = 2.0 + cf_vector.at(1) / cf_vector.at(0);
+                double c = 1.0 + cf_vector.at(1) / cf_vector.at(0) + cf_vector.at(2) / cf_vector.at(0);
+                initial_guess = -0.5 * b - 0.5 * sqrt(b * b - 4.0 * c);
+                if ((initial_guess <= 0) || (initial_guess >= 1)) initial_guess = -0.5 * b + 0.5 * sqrt(b * b - 4.0 * c);
+            }
+        }
+        else if (initial_guess < 0) // first order
+        {
+            if (cf_vector.at(0) != 0) initial_guess = -(1.0 + cf_vector.at(1) / cf_vector.at(0));
+        }
+
+        double scale_factor = irr_scale_factor(cf_vector, count);
+        double residual = DBL_MAX;
```

Comment:
DBL_MAX was not defined and is causing the Linux and MacOS tests to fail.

DBL_MAX is defined in float.h and was previously included at the top of each financial compute module as
#ifndef WIN32
#include <float.h>
#endif